ml_genn/ml_genn/compilers/event_prop_compiler.py

```diff
@@ -0,0 +1,783 @@
+    $(addToInSyn, $(g));
+    """,
+    "event_threshold_condition_code": """
+    $(BackSpike_pre)
+    """,
+    "event_code": """
+    $(Gradient) -= ($(LambdaI_post) * $(TauSyn));
+    """}
+
+# EventProp weight update model used on KERNEL weights when atomics required
+# **NOTE** feedback is added if required
+# **YUCK** you should NEVER have to use backend-specific code in models
+weight_update_model_atomic_cuda = {
+    "param_name_types": [("TauSyn", "scalar")],
+    "var_name_types": [("g", "scalar", VarAccess_READ_ONLY),
+                       ("Gradient", "scalar")],
+
+    "sim_code": """
+    $(addToInSyn, $(g));
+    """,
+    "event_threshold_condition_code": """
+    $(BackSpike_pre)
+    """,
+    "event_code": """
+    atomicAdd(&$(Gradient), -($(LambdaI_post) * $(TauSyn)));
+    """}
+
+# Weight update model used on non-trainable connections
+# **NOTE** feedback is added if required
+static_weight_update_model = {
+    "param_name_types": [("g", "scalar")],
+    "sim_code":
+        """
+        $(addToInSyn, $(g));
+        """,
+    "event_threshold_condition_code": """
+    $(BackSpike_pre)
+    """,}
+
+gradient_batch_reduce_model = {
+    "var_name_types": [("ReducedGradient", "scalar", VarAccess_REDUCE_BATCH_SUM)],
+    "var_refs": [("Gradient", "scalar")],
+    "update_code": """
+    $(ReducedGradient) = $(Gradient);
+    $(Gradient) = 0;
+    """}
+
+# Template used to generate backward passes for neurons
+neuron_backward_pass = Template(
+    """
+    const int ringOffset = ($$(batch) * $$(num) * $max_spikes) + ($$(id) * $max_spikes);
+    const scalar backT = $example_time - $$(t) - DT;
+
+    // Backward pass
+    $dynamics
+    if ($$(BackSpike)) {
+        $transition
+
+        // Decrease read pointer
+        $$(RingReadOffset)--;
+        if ($$(RingReadOffset) < 0) {
+            $$(RingReadOffset) = $max_spikes - 1;
+        }
+        $$(BackSpike) = false;
+    }
+    // YUCK - need to trigger the back_spike the time step before to get the correct backward synaptic input
+    if ($$(RingReadOffset) != $$(RingReadEndOffset) && fabs(backT - $$(RingSpikeTime)[ringOffset + $$(RingReadOffset)] - DT) < 1e-3*DT) {
+        $$(BackSpike) = true;
+    }
+
+    // Forward pass
+    """)
+
+# Template used to generate reset code for neurons
+neuron_reset = Template(
+    """
+    if($$(RingWriteOffset) != $$(RingReadEndOffset)) {
+        // Write spike time and I-V to tape
+        $$(RingSpikeTime)[ringOffset + $$(RingWriteOffset)] = $$(t);
+        $write
+        $$(RingWriteOffset)++;
+        
+        // Loop around if we've reached end of circular buffer
+        if ($$(RingWriteOffset) >= $max_spikes) {
+            $$(RingWriteOffset) = 0;
+        }
+    }
+    else {
+        //printf("%f: hidden: ImV buffer violation in neuron %d, fwd_start: %d, new_fwd_start: %d, rp_ImV: %d, wp_ImV: %d\\n", $$(t), $$(id), $$(RingReadEndOffset), $$(RingWriteStartOffset), $$(RingReadOffset), $$(RingWriteOffset));
```

Comment:
I see you left this here - what is your opinion about offering some sort of "strict buffer checking" flag that would enable this code?

Reply:
Good idea - have added a ``strict_buffer_checking`` flag

Reply:
prefect!